src/tasks/gitea/gitea-drone-oauth.ts

```diff
@@ -0,0 +1,225 @@
+      password: env.GITEA_PASSWORD,
+    },
+  }
+
+  console.log('Authorizing OAuth application')
+
+  const authorizeResponse: AxiosResponse<any> = await axios.get(`${giteaUrl}/login/oauth/authorize`, options)
+
+  return authorizeResponse.headers['set-cookie']
+}
+
+export function parseCSRFToken(authorizeHeaderCookies: string[]): string {
+  // Loop over cookies and find the _csrf cookie and retrieve the value
+  const csrfCookies = authorizeHeaderCookies.filter((cookie) => {
+    return cookie.includes(csrfCookieName) // Find cookie that contains _csrf
+  })
+
+  if (csrfCookies.length == 0) {
+    throw new GiteaDroneError('No CSRF cookie was returned')
+  }
+
+  const csrfTokens = csrfCookies[0]
+    .split(';') // Split that cookie into multiple parts, as a single cookie can contain multiple key/values ';' separated
+    .map((c) => c.trim())
+    .filter((cookie) => {
+      return cookie.includes(csrfCookieName) // Find the key/value pair with _csrf as key
+    })
+    .map((cookie) => {
+      return cookie.substring(csrfCookieName.length + 1) // Retrieve value for '_csrf'-key
+    })
+
+  if (csrfTokens.length == 0 || csrfTokens[0].length == 0) {
+    throw new GiteaDroneError('No CSRF token was returned')
+  }
+  return csrfTokens[0]
+}
+async function authorizeOAuthApp(giteaUrl: string, droneLoginUrl: string, oauthData: DroneSecret) {
+  let authorizeHeaderCookies: string[]
+  try {
+    authorizeHeaderCookies = await getGiteaAuthorizationHeaderCookie(giteaUrl, droneLoginUrl, oauthData)
+  } catch (error) {
+    throw new GiteaDroneError('Authorization already granted or something went wrong')
+  }
+
+  const csrfToken = parseCSRFToken(authorizeHeaderCookies)
+
+  console.log('Granting authorization')
+
+  const grantOptions: AxiosRequestConfig = {
+    method: 'POST',
+    url: `${giteaUrl}/login/oauth/grant`,
+    headers: {
+      cookie: authorizeHeaderCookies
+        .map((cookie) => {
+          return cookie.split(';')[0]
+        })
+        .join('; '),
+      Cookie: authorizeHeaderCookies
+        .map((cookie) => {
+          return cookie.split(';')[0]
+        })
+        .join('; '),
+    },
+    maxRedirects: 1,
+    auth: {
+      username: env.GITEA_USER,
+      password: env.GITEA_PASSWORD,
+    },
+    // Data for this post query must be stringified https://github.com/axios/axios#using-applicationx-www-form-urlencoded-format
+    data: querystring.stringify({
+      _csrf: csrfToken,
+      // eslint-disable-next-line @typescript-eslint/camelcase
+      client_id: `${oauthData.clientId}`,
+      // eslint-disable-next-line @typescript-eslint/camelcase
+      redirect_uri: droneLoginUrl,
+    }),
+  }
+
+  try {
+    await axios.request(grantOptions)
+  } catch (error) {
+    // Do nothing, error code could be on the redirect
+  }
+}
+
+export function isSecretValid(remoteSecret: DroneSecret, oauthApps: OAuth2Application[]): boolean {
+  if (!remoteSecret) {
+    console.log('Remote secret was not found')
```

Comment:
I understand what you're saying, let me try to explain my reasoning.
I could run this if/elseif/else combination within the `main` function, but that would increase the codebase in that function aka: clutter.
By separating this into a separate function, it also becomes testable.

You say: `it has not "searched" for anything` and within the context of only this function you are correct. But this function is called after the (now renamed to) `getSecret` function, that passes it's output to this function as a `remoteSecret`

Making it an interface makes `null` still valid under current linting conditions (strict is false btw)
I think making it strict, should be another issue, as this could pose problems for other tasks as well.

Reply:
You can make it strict by using an interface or typing.

Reply:
Why is null valid?

Reply:
Because tsconfig.json sets strict to false. And `strictNullCheck` is not set
https://www.typescriptlang.org/tsconfig#strictNullChecks